keps/sig-node/127-usernamespaces-support/README.md

```diff
@@ -0,0 +1,842 @@
+  - https://github.com/opencontainers/runc/commit/0a8e4117e7f715d5fbeef398405813ce8e88558b
+
+### Goals
+
+- Increase node to pod isolation in Kubernetes by mapping user and group IDs
+  inside the container to different IDs in the host. In particular, mapping root
+  inside the container to unprivileged user and group IDs in the node.
+- Make it safer to run workloads that need highly privileged capabilities such
+  as `CAP_SYS_ADMIN`, reducing the risk of impacting the host.
+- Benefit from the security hardening that user namespaces provide against some
+  of the future unknown runtime and kernel vulnerabilities.
+
+### Non-Goals
+
+- Provide a way to run the kubelet process or container runtimes as an
+  unprivileged process. Although initiatives like
+  [usernetes](https://github.com/rootless-containers/usernetes) and this KEP
+  both make use of user namespaces, it is a different implementation for a
+  different purpose.
+
+## Proposal
+
+This proposal aims to support user namespaces in Kubernetes by extending the pod
+specification with a new `userNamespaceMode` field. This field can have 3
+values:
+
+- **Host**:
+  The pods are placed in the host user namespace, this is the current Kubernetes
+  behaviour. This mode is intended for pods that only work in the initial (host)
+  user namespace. It is the default mode when `userNamespaceMode` field is not
+  set.
+
+- **Cluster**:
+  All the pods in the cluster are placed in a _unique_ user namespace but they
+  use the same ID mappings. This mode doesn't provide full pod-to-pod isolation
+  as all the pods with `Cluster` mode have the same effective IDs on the host.
+  It provides pod-to-host isolation as the IDs are different inside the
+  container and in the host. This mode is intended for pods sharing volumes as
+  they will run with the same effective IDs on the host, allowing them to read
+  and write files in the shared storage.
+
+- **Pod**:
+  Each pod is placed in a different user namespace and has a different and
+  non-overlapping ID mapping across the node. This mode is intended for stateless pods, i.e.
+  pods using only ephemeral volumes like `configMap,` `downwardAPI`, `secret`,
+  `projected` and `emptyDir`. This mode not only provides host-to-pod isolation
+  but also pod-to-pod isolation as each pod has a different range of effective
+  IDs in the host.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I want run some pods with privileged capabilities because
+the applications in the pods require it (e.g. `CAP_SYS_ADMIN` to mount a FUSE
+filesystem or `CAP_NET_ADMIN` to setup a VPN) but I don't want this extra
+capability to grant me any extra privilege on the host.
+
+#### Story 2
+
+As a cluster admin, I want to allow some pods to run in the host user namespace
+if they need a feature only available in such user namespace, such as loading a
+kernel module with `CAP_SYS_MODULE`.
+
+### Notes/Constraints/Caveats
```

Comment:
Have you thought about devices' permissions (device mounts are set by device plugins in kubernetes)? AFAICS the problem space is similar to volumes but devices don't have the luxury of `fsGroups` we could use for GID mapping. The additional challenge with devices is that different devices have their own GIDs and we've also seen that certain groups (e.g., `video`) have different GID value depending on the node OS.

kubernetes/issues/92211 talks about the challenges with non-root users we've tried to solve. The primary goal is to allow containers with non-root user processes access devices. It looks at least `Host` mode is compatible with that. User namespaces was recently brought up and thus my questions here.